lib/SPIRV/libSPIRV/SPIRVInstruction.h

```diff
@@ -2757,11 +2757,28 @@ _SPIRV_OP(ImageQuerySamples, true, 4)
 #define _SPIRV_OP(x, ...)                                                      \
   typedef SPIRVInstTemplate<SPIRVInstTemplateBase, Op##x, __VA_ARGS__> SPIRV##x;
 // Other instructions
-_SPIRV_OP(SpecConstantOp, true, 4, true, 0)
 _SPIRV_OP(GenericPtrMemSemantics, true, 4, false)
 _SPIRV_OP(GenericCastToPtrExplicit, true, 5, false, 1)
 #undef _SPIRV_OP
 
+class SPIRVSpecConstantOpBase : public SPIRVInstTemplateBase {
+public:
+  std::vector<SPIRVEntry *> getNonLiteralOperands() const override {
+    // If SpecConstant results from CompositeExtract/Insert operation, then all
+    // operands are expected to be literals.
+    SPIRVWord LiteralOpcode = Ops[0];
+    if (LiteralOpcode == OpCompositeExtract ||
+        LiteralOpcode == OpCompositeInsert)
+      return {};
+
+    return SPIRVInstTemplateBase::getNonLiteralOperands();
+  }
```

Comment:
Still, we may call `isOperandLiteral()` which will return wrong answer. I suggest overriding `isOperandLiteral()` instead
```suggestion
  bool isOperandLiteral(unsigned I) const override {
    switch (Ops[0]) { // Opcode
    case OpCompositeExtract:
    case OpCompositeInsert:
      return true;
    default:
      return SPIRVInstTemplateBase::isOperandLiteral(I); // Will it always return false?
    }
  }
```

Reply:
Nice suggestion, thanks! Not sure if `SPIRVInstTemplateBase::isOperandLiteral(I)` will always return `false`, but it looks right to call the base class method to be safe for sure.